File: src/domain/cache/coverage.ts
/**
 * Coverage statistics domain entities
 *
 * Defines entities for tracking audio cache coverage per document.
 */

/** Coverage statistics for a document's audio cache */
export interface CoverageStats {
  documentId: string;
  totalChunks: number;
  cachedChunks: number;
  coveragePercent: number;
  totalDurationMs: number;
  cachedSizeBytes: number;
  lastUpdated: string;
  pageStats?: PageCoverageStats[];
}

/** Coverage statistics for a single page */
export interface PageCoverageStats {
  pageNumber: number;
  totalChunks: number;
  cachedChunks: number;
  coveragePercent: number;
  durationMs: number;
}

/**
 * Create empty coverage stats for a document
 */
export function emptyCoverageStats(documentId: string): CoverageStats {
  return {
    documentId,
    totalChunks: 0,
    cachedChunks: 0,
    coveragePercent: 0,
    totalDurationMs: 0,
    cachedSizeBytes: 0,
    lastUpdated: new Date().toISOString(),
  };
}

/**
 * Calculate coverage percentage from counts
 */
export function calculateCoveragePercent(
  cached: number,
  total: number,
): number {
  if (total === 0) return 0;
  return Math.round((cached / total) * 100);
}

/**
 * Check if a document is fully cached
 */
export function isFullyCached(stats: CoverageStats): boolean {
  return stats.cachedChunks === stats.totalChunks && stats.totalChunks > 0;
}

/**
 * Format coverage as a human-readable string
 */
export function formatCoverage(stats: CoverageStats): string {
  if (stats.totalChunks === 0) {
    return "No audio cached";
  }
  return `${stats.coveragePercent}% cached (${stats.cachedChunks}/${stats.totalChunks} chunks)`;
}

/**
 * Format duration as human-readable string
 */
export function formatDuration(durationMs: number): string {
  if (durationMs < 1000) {
    return `${durationMs}ms`;
  }

  const seconds = Math.floor(durationMs / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);

  if (hours > 0) {
    const remainingMinutes = minutes % 60;
    return `${hours}h ${remainingMinutes}m`;
  }

  if (minutes > 0) {
    const remainingSeconds = seconds % 60;
    return `${minutes}m ${remainingSeconds}s`;
  }

  return `${seconds}s`;
}

/**
 * Format bytes as human-readable size
 */
export function formatBytes(bytes: number): string {
  if (bytes === 0) return "0 B";

  const units = ["B", "KB", "MB", "GB", "TB"];
  const k = 1024;
  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${units[i]}`;
}

File: src-tauri/src/domain/cache/coverage.rs
//! Coverage statistics domain entities
//!
//! Defines entities for tracking audio cache coverage per document.

use serde::{Deserialize, Serialize};

/// Coverage statistics for a document's audio cache
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CoverageStats {
    pub document_id: String,
    pub total_chunks: i32,
    pub cached_chunks: i32,
    pub coverage_percent: f64,
    pub total_duration_ms: i64,
    pub cached_size_bytes: i64,
    pub last_updated: String,
    pub page_stats: Option<Vec<PageCoverageStats>>,
}

impl CoverageStats {
    /// Create empty coverage stats for a document
    pub fn empty(document_id: String) -> Self {
        Self {
            document_id,
            total_chunks: 0,
            cached_chunks: 0,
            coverage_percent: 0.0,
            total_duration_ms: 0,
            cached_size_bytes: 0,
            last_updated: chrono::Utc::now().to_rfc3339(),
            page_stats: None,
        }
    }

    /// Check if the document is fully cached
    pub fn is_fully_cached(&self) -> bool {
        self.cached_chunks == self.total_chunks && self.total_chunks > 0
    }

    /// Calculate percentage from counts
    pub fn calculate_percent(cached: i32, total: i32) -> f64 {
        if total == 0 {
            return 0.0;
        }
        ((cached as f64 / total as f64) * 100.0).round()
    }

    /// Update coverage stats with new data
    pub fn update(
        &mut self,
        total_chunks: i32,
        cached_chunks: i32,
        total_duration_ms: i64,
        cached_size_bytes: i64,
    ) {
        self.total_chunks = total_chunks;
        self.cached_chunks = cached_chunks;
        self.coverage_percent = Self::calculate_percent(cached_chunks, total_chunks);
        self.total_duration_ms = total_duration_ms;
        self.cached_size_bytes = cached_size_bytes;
        self.last_updated = chrono::Utc::now().to_rfc3339();
    }
}

/// Coverage statistics for a single page
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PageCoverageStats {
    pub page_number: i32,
    pub total_chunks: i32,
    pub cached_chunks: i32,
    pub coverage_percent: f64,
    pub duration_ms: i64,
}

impl PageCoverageStats {
    /// Create new page coverage stats
    pub fn new(page_number: i32, total_chunks: i32, cached_chunks: i32, duration_ms: i64) -> Self {
        Self {
            page_number,
            total_chunks,
            cached_chunks,
            coverage_percent: CoverageStats::calculate_percent(cached_chunks, total_chunks),
            duration_ms,
        }
    }

    /// Check if the page is fully cached
    pub fn is_fully_cached(&self) -> bool {
        self.cached_chunks == self.total_chunks && self.total_chunks > 0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_coverage() {
        let stats = CoverageStats::empty("doc-1".to_string());
        assert_eq!(stats.document_id, "doc-1");
        assert_eq!(stats.total_chunks, 0);
        assert_eq!(stats.cached_chunks, 0);
        assert_eq!(stats.coverage_percent, 0.0);
        assert!(!stats.is_fully_cached());
    }

    #[test]
    fn test_calculate_percent() {
        assert_eq!(CoverageStats::calculate_percent(0, 0), 0.0);
        assert_eq!(CoverageStats::calculate_percent(0, 10), 0.0);
        assert_eq!(CoverageStats::calculate_percent(5, 10), 50.0);
        assert_eq!(CoverageStats::calculate_percent(10, 10), 100.0);
        assert_eq!(CoverageStats::calculate_percent(1, 3), 33.0); // Rounded
    }

    #[test]
    fn test_is_fully_cached() {
        let mut stats = CoverageStats::empty("doc-1".to_string());

        // Empty is not fully cached
        assert!(!stats.is_fully_cached());

        // Partial coverage
        stats.total_chunks = 10;
        stats.cached_chunks = 5;
        assert!(!stats.is_fully_cached());

        // Full coverage
        stats.cached_chunks = 10;
        assert!(stats.is_fully_cached());
    }

    #[test]
    fn test_update_coverage() {
        let mut stats = CoverageStats::empty("doc-1".to_string());
        stats.update(100, 75, 300000, 1024000);

        assert_eq!(stats.total_chunks, 100);
        assert_eq!(stats.cached_chunks, 75);
        assert_eq!(stats.coverage_percent, 75.0);
        assert_eq!(stats.total_duration_ms, 300000);
        assert_eq!(stats.cached_size_bytes, 1024000);
    }

    #[test]
    fn test_page_coverage_stats() {
        let page = PageCoverageStats::new(1, 10, 8, 60000);

        assert_eq!(page.page_number, 1);
        assert_eq!(page.total_chunks, 10);
        assert_eq!(page.cached_chunks, 8);
        assert_eq!(page.coverage_percent, 80.0);
        assert_eq!(page.duration_ms, 60000);
        assert!(!page.is_fully_cached());
    }

    #[test]
    fn test_page_fully_cached() {
        let page = PageCoverageStats::new(1, 5, 5, 30000);
        assert!(page.is_fully_cached());
    }
}
